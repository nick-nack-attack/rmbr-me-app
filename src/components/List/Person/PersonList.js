// component displays list of people user has
import React, { Component, useContext, useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';

// context
import { AppContext } from '../../../contexts/AppContext';

// service / util
import AppApiService from '../../../services/app-api-service';
import { findRmbrByPersonId } from "../../Utils/helpers";

// component
import AddPersonForm from '../../Form/Person/Add/AddPersonForm';
import PersonItem from "../../Item/Person/PersonItem";

// style
import emptyStateImage from "../../../assets/empty_people.png"
import "../List.scss";

// components

const PersonList = () => {

    // set context
    let appContext = useContext(AppContext);
    const history = useHistory();

    const [ people, setPeople ] = useState([]);
    const [ rmbrs, setRmbrs ] = useState([]);
    const [ searchTerm, setSearchTerm ] = useState('');
    const [ results, setResults ] = useState([]);

    const handleClick = (id) => {
        history.push(`/person/${id}`);
    };

    const handleChange = e => {
        setSearchTerm(e);
    };

    useEffect(() => {
        setPeople(appContext.state.people);
        setRmbrs(appContext.state.rmbrs);
    }, [ appContext ]);

    useEffect(() => {
        const filterBySearch = people.filter(r => r.person_name.toLowerCase().includes(searchTerm.toLowerCase()));
        setResults(filterBySearch);
    }, [ searchTerm ]);


    const renderemptystate = (
        <div>
            <img
                src={ emptyStateImage }
                alt="lightning says he's lonely"
            />
            <div>
                <p>Add a person to start!</p>
            </div>
        </div>
    );    

    const renderPeople = () => {
        return (
            <ul> 
                {
                    people.sort((a,b) => b.id - a.id).map(person => {
                        return (
                            <PersonItem 
                                person={person} 
                                // pass in only the rmbrs the person has
                                array={ findRmbrByPersonId(rmbrs, person.id) }
                                handleClickPerson={ id => handleClick(id) }
                            />
                        )
                    })
                }   
            </ul>
        );
    };

    const renderAnnoucement = () => {
        return (
            <h2>
                You have { people.length } with { rmbrs.length } rmbrs.
            </h2>
        )
    }

    return (

        <>

        <AddPersonForm />

        { renderAnnoucement() }
        
        { people === undefined ? renderemptystate : '' }

        { renderPeople() }

        </>

    );

};

export default PersonList;