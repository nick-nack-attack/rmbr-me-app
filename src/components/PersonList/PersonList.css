
.person-cell-button-bar > button {
    margin-right: 10px;
}
