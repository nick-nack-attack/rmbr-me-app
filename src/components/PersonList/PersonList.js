import React, { Component } from 'react';
import PersonContext from "../../contexts/PersonContext";
import RmbrApiService from '../../services/rmbr-api-service';
import { Section } from "../../components/Utils/Utils";
import PersonListItem from "../../components/PersonListItem/PersonListItem";
import AddPersonForm from '../../components/AddPersonForm/AddPersonForm'
import { findRmbrByPersonId } from "../../helpers";

export default class PersonList extends Component {

    static contextType = PersonContext;

    componentDidMount() {
        this.context.clearError()
        const user_id = window.localStorage.getItem('user_id');
        RmbrApiService.getPersonByUserId(user_id)
            .then(this.context.setPersonArray)
            .catch(this.context.setError)
        RmbrApiService.getRmbrByUserId(user_id)
            .then(res => this.context.setRmbrArray(res))
            .catch(this.context.setError)
    };

    renderPeople() {
        const { personArray = [] } = this.context
        const { rmbrArray = [] } = this.context
        return personArray.map(person =>
            <PersonListItem
                person={person}
                rmbrArray={findRmbrByPersonId(rmbrArray, person.id)}
            />
        )
    };

    render() {

        const { error } = this.context;

        return (
            <>
                <ul id='person_list'>
                    <label>You have <b>{this.context.personArray.length}</b> people with <b>{this.context.rmbrArray.length}</b> rmbrs</label>
                    { error
                        ? <p className='red'>There was an error, try again</p>
                        : this.renderPeople() }
                </ul>

                <div className='PeopleListPage__add-person-button'>
                    <AddPersonForm/>
                </div>
            </>
        )
    }
}


