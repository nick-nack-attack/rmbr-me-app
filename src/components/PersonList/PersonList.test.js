import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import PersonList from "./PersonList";

describe(`PersonList component`, () => {

    it('renders PersonList by default', () => {
        const wrapper = shallow(<PersonList/>)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

})