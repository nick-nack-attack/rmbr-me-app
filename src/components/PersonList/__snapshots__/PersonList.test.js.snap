// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonList component renders PersonList by default 1`] = `
<testContext>
  <PersonList
    context={
      Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "personArray": Array [],
        },
        "_currentValue2": Object {
          "personArray": Array [],
        },
        "_threadCount": 0,
      }
    }
  />
</testContext>
`;
